crates/reward/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+    fn inner_to_balance(x: Inner<T, I>) -> Result<T::Balance, DispatchError> {
+        let y = UniqueSaturatedInto::<u128>::unique_saturated_into(x);
+        TryInto::<T::Balance>::try_into(y).map_err(|_| Error::<T, I>::TryIntoIntError.into())
+    }
+
+    fn balance_to_inner(x: T::Balance) -> Result<Inner<T, I>, DispatchError> {
+        let y = TryInto::<u128>::try_into(x).map_err(|_| Error::<T, I>::TryIntoIntError)?;
+        TryInto::<Inner<T, I>>::try_into(y).map_err(|_| Error::<T, I>::TryIntoIntError.into())
+    }
+
+    fn balance_to_fixed(x: T::Balance) -> Result<SignedFixedPoint<T, I>, DispatchError> {
+        let unsigned_fixed_point = SignedFixedPoint::<T, I>::checked_from_integer(Self::balance_to_inner(x)?)
+            .ok_or(Error::<T, I>::TryIntoIntError)?;
+        Ok(unsigned_fixed_point)
+    }
+
+    pub fn participants() -> Vec<(T::AccountId, T::Balance)> {
+        <Stake<T, I>>::iter().collect()
+    }
+}
+
+pub trait Rewards<AccountId> {
+    /// The balance type for a currency.
+    type Balance: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default;
+
+    /// Return the stake associated with the `account_id`.
+    fn get_stake(account_id: &AccountId) -> Self::Balance;
+
+    /// Deposit an `amount` of stake to the `account_id`.
+    fn deposit_stake(account_id: &AccountId, amount: Self::Balance) -> Result<(), DispatchError>;
+
+    /// Distribute the `reward` to all participants.
+    fn distribute(reward: Self::Balance) -> Result<Self::Balance, DispatchError>;
+
+    /// Compute the expected reward for the `account_id`.
+    fn compute_reward(account_id: &AccountId) -> Result<Self::Balance, DispatchError>;
+
+    /// Withdraw an `amount` of stake from the `account_id`.
+    fn withdraw_stake(account_id: &AccountId, amount: Self::Balance) -> Result<(), DispatchError>;
+
+    /// Withdraw all rewards from the `account_id`.
+    fn withdraw_reward(account_id: &AccountId) -> Result<Self::Balance, DispatchError>;
+}
+
+impl<T: Config<I>, I: 'static> Rewards<T::AccountId> for Pallet<T, I>
+where
+    T::Balance: MaybeSerializeDeserialize + Debug,
+{
+    type Balance = T::Balance;
+
+    fn get_stake(account_id: &T::AccountId) -> Self::Balance {
+        Self::stake(account_id)
+    }
+
+    fn deposit_stake(account_id: &T::AccountId, amount: Self::Balance) -> Result<(), DispatchError> {
+        <TotalStake<T, I>>::mutate(|total_stake| {
+            *total_stake = total_stake
+                .checked_add(&amount)
+                .ok_or(Error::<T, I>::ArithmeticOverflow)?;
+            Ok::<_, Error<T, I>>(())
+        })?;
+
+        <Stake<T, I>>::mutate(account_id, |stake| {
+            *stake = stake.checked_add(&amount).ok_or(Error::<T, I>::ArithmeticOverflow)?;
+            Ok::<_, Error<T, I>>(())
+        })?;
+
+        let amount_as_fixed = Self::balance_to_fixed(amount)?;
+        <RewardTally<T, I>>::mutate(account_id, |reward_tally| {
```

Comment:
Do we need to worry about this overflowing?

Reply:
I don't think so, it would take a lot of rewards and a very high total stake.